### PR TITLE
Extend and refactor waveform metrics

### LINKED ISSUE #1185: [feat-postprocessing]: implement missing template metrics
While most of the template/waveform metrics from the original [ecephys pipeline](https://github.com/AllenInstitute/ecephys_spike_sorting/tree/master/ecephys_spike_sorting/modules/mean_waveforms) are implemented via: `si.calculate_template_metrics(we)`, a 1D (and all 2D) metrics are missing including:

1D:
- [x] duration

2D:
- [ ] Waveform spread
- [ ] velocity above the soma
- [ ] velocity below the soma

### PR DESCRIPTION
**Pull request by alejoe91**
Implemented additional metrics:

### Single-channel

- num_positive_peaks
- num_negative_peaks

### Multi-channel

- velocity_above
- velocity_below
- spread
- exp_decay

Closes  #1185

### CODE REVIEW

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 1]
```diff
@@ -100,15 +132,50 @@ def _run(self):
...
+                    sampling_frequency_up = sampling_frequency
+
+                func = _metric_name_to_func[metric_name]
+                value = func(
+                    template_upsampled,
+                    channel_locations=channel_locations_sparse,
+                    sampling_frequency=sampling_frequency_up,
+                    **self._params["functions_kwargs"],
+                )
+                template_metrics.at[index, metric_name] = value
```
Comment (samuelgarcia) [inline src/spikeinterface/postprocessing/template_metrics.py:178]: Is value a scalr or a vector ?

Comment (alejoe91) [reply src/spikeinterface/postprocessing/template_metrics.py:175]: scalar always

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 2]
```diff
@@ -197,49 +329,106 @@ def get_trough_and_peak_idx(template):
     """
     Return the indices into the input template of the detected trough
     (minimum of template) and peak (maximum of template, after trough).
-    Assumes negative trough and positive peak
+    Assumes negative trough and positive peak.
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:332]: I'm curious about the use of the word peak in this context. Typically we compare trough (min) vs crest (max). In SpikeInterface the user usually has the choice to specify the peak as either 'pos' or 'neg'. So would it be clearer to use the trough and crest to be absolute terms and the user decide on the meaning of peak based on `peak_sign` type arguments in general? This isn't enforced in si, but I'm more just curious.

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 3]
```diff
@@ -197,49 +329,106 @@ def get_trough_and_peak_idx(template):
...
+    Parameters
+    ----------
+    template: numpy.ndarray
+        The 1D template waveform
+
+    Returns
+    -------
+    trough_idx: int
+        The index of the trough
+    peak_idx: int
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:343]: For example if we call this crest instead then we don't worry about the fact that for many extracellular recordings we want to define the peak as the most negative voltage?

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 4]
```diff
@@ -197,49 +329,106 @@ def get_trough_and_peak_idx(template):
...
-    Time between trough and peak in s
+    Return the peak to valley duration in seconds of input waveforms.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    trough_idx: int, default: None
+        The index of the trough
+    peak_idx: int, default: None
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:364]:
```suggestion
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    trough_idx: int, default: None
        The index of the trough
    peak_idx: int, default: None
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 5]
```diff
@@ -197,49 +329,106 @@ def get_trough_and_peak_idx(template):
...
+    Return the peak to trough ratio of input waveforms.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    trough_idx: int, default: None
+        The index of the trough
+    peak_idx: int, default: None
+        The index of the peak
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:389]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float or None, default: None
        The sampling frequency for the recording
    trough_idx: int, default: None
        The index of the trough
    peak_idx: int, default: None
        The index of the peak
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 6]
```diff
@@ -197,49 +329,106 @@ def get_trough_and_peak_idx(template):
...
+    Return the half width of input waveforms in seconds.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    trough_idx: int, default: None
+        The index of the trough
+    peak_idx: int, default: None
+        The index of the peak
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:413]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    trough_idx: int, default: None
        The index of the trough
    peak_idx: int, default: None
        The index of the peak
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 7]
```diff
@@ -264,17 +453,23 @@ def get_repolarization_slope(template, **kwargs):
...
-    """
 
-    trough_idx, peak_idx = get_trough_and_peak_idx(template)
-    sampling_frequency = kwargs["sampling_frequency"]
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    trough_idx: int, default: None
+        The index of the trough
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:462]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    trough_idx: int, default: None
        The index of the trough
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 8]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
     recovery slopes per waveform.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    peak_idx: int, default: None
+        The index of the peak
+    **kwargs: Required kwargs:
+        - recovery_window_ms: the window in ms after the peak to compute the recovery_slope
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:505]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    peak_idx: int, default: None
        The index of the peak
    **kwargs: Required kwargs:
        - recovery_window_ms: the window in ms after the peak to compute the recovery_slope
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 9]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    """
+    Count the number of positive peaks in the template.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    **kwargs: Required kwargs:
+        - peak_relative_threshold: the relative threshold to detect positive and negative peaks
+        - peak_width_ms: the width in samples to detect peaks
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:535]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    **kwargs: Required kwargs:
        - peak_relative_threshold: the relative threshold to detect positive and negative peaks
        - peak_width_ms: the width in samples to detect peaks
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 10]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    """
+    Count the number of negative peaks in the template.
+
+    Parameters
+    ----------
+    template_single: numpy.ndarray
+        The 1D template waveform
+    **kwargs: Required kwargs:
+        - peak_relative_threshold: the relative threshold to detect positive and negative peaks
+        - peak_width_ms: the width in samples to detect peaks
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:561]:
```suggestion
    template_single: numpy.ndarray
        The 1D template waveform
    sampling_frequency : float
        The sampling frequency for the recording
    **kwargs: Required kwargs:
        - peak_relative_threshold: the relative threshold to detect positive and negative peaks
        - peak_width_ms: the width in samples to detect peaks
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 11]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    Compute the velocity above the max channel of the template.
+
+    Parameters
+    ----------
+    template: numpy.ndarray
+        The template waveform (num_samples, num_channels)
+    channel_locations: numpy.ndarray
+        The channel locations (num_channels, 2)
+    **kwargs: Required kwargs:
+        - depth_direction: the direction to compute velocity above and below ("x", "y", or "z")
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:636]:
```suggestion
    channel_locations: numpy.ndarray
        The channel locations (num_channels, 2)
    sampling_frequency : float
        The sampling frequency for the recording
    **kwargs: Required kwargs:
        - depth_direction: the direction to compute velocity above and below ("x", "y", or "z")
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 12]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    """
+    Compute the velocity below the max channel of the template.
+
+    Parameters
+    ----------
+    template: numpy.ndarray
+        The template waveform (num_samples, num_channels)
+    channel_locations: numpy.ndarray
+        The channel locations (num_channels, 2)
+    **kwargs: Required kwargs:
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:710]:
```suggestion
    channel_locations: numpy.ndarray
        The channel locations (num_channels, 2)
    sampling_frequency : float
        The sampling frequency for the recording
    **kwargs: Required kwargs:
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 13]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    """
+    Compute the exponential decay of the template amplitude over distance.
+
+    Parameters
+    ----------
+    template: numpy.ndarray
+        The template waveform (num_samples, num_channels)
+    channel_locations: numpy.ndarray
+        The channel locations (num_channels, 2)
+    **kwargs: Required kwargs:
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:784]:
```suggestion
    channel_locations: numpy.ndarray
        The channel locations (num_channels, 2)
    sampling_frequency : float or None, default: None
        The sampling frequency for the recording
    **kwargs: Required kwargs:
```

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 14]
```diff
@@ -299,41 +494,427 @@ def get_recovery_slope(template, window_ms=0.7, **kwargs):
...
+    Compute the spread of the template amplitude over distance.
+
+    Parameters
+    ----------
+    template: numpy.ndarray
+        The template waveform (num_samples, num_channels)
+    channel_locations: numpy.ndarray
+        The channel locations (num_channels, 2)
+    **kwargs: Required kwargs:
+        - depth_direction: the direction to compute velocity above and below ("x", "y", or "z")
```
Comment (zm711) [inline src/spikeinterface/postprocessing/template_metrics.py:860]:
```suggestion
    channel_locations: numpy.ndarray
        The channel locations (num_channels, 2)
    sampling_frequency : float
        The sampling frequency for the recording
    **kwargs: Required kwargs:
        - depth_direction: the direction to compute velocity above and below ("x", "y", or "z")
```

**Review by zm711 — COMMENTED** [review]
I did a partial update of some of the docstrings (a few typos and added in the sampling_frequency parameter into the parameters section, but maybe there was a reason why you didn't want them so feel free to delete those they are not needed.

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 15]
```diff
@@ -157,34 +246,77 @@ def compute_template_metrics(
...
-    upsampling_factor : int, optional
-        Upsample factor, by default 10
-    sparsity: dict or None
+    peak_sign : {"neg", "pos"}, default: "neg"
+        The peak sign
+    upsampling_factor : int, default: 10
+        The upsampling factor to upsample the templates
+    sparsity: dict or None, default: None
         Default is sparsity=None and template metric is computed on extremum channel only.
         If given, the dictionary should contain a unit ids as keys and a channel id or a list of channel ids as values.
```
[the remark this answers is not in the document]
Comment (alejoe91) [reply src/spikeinterface/postprocessing/template_metrics.py]: updated with proper ChannelSparsity docstring

**Comment by alejoe91** [comment]
Thanks @zm711 

Aplied the suggested changes. @samuelgarcia @h-mayorquin good to merge

File: src/spikeinterface/core/waveform_extractor.py [hunk 16]
```diff
@@ -2016,6 +2032,9 @@ def set_params(self, **params):
         params = self._set_params(**params)
         self._params = params
 
+        if self.waveform_extractor.is_read_only():
```
Comment (samuelgarcia) [inline src/spikeinterface/core/waveform_extractor.py:2035]: Maybe we should run error instead no ?
set_params on read only should be allowed.

File: src/spikeinterface/postprocessing/template_metrics.py [hunk 17]
```diff
@@ -132,14 +197,31 @@ def get_extension_function():
...
-    sparsity=None,
-    window_slope_ms=0.7,
+    load_if_exists: bool = False,
+    metric_names: Optional[list[str]] = None,
+    peak_sign: Optional[str] = "neg",
+    upsampling_factor: int = 10,
+    sparsity: Optional[ChannelSparsity] = None,
+    include_multi_channel_metrics: bool = False,
+    metrics_kwargs: dict = None,
+    debug_plots: bool = False,
```
Comment (samuelgarcia) [inline src/spikeinterface/postprocessing/template_metrics.py:224]: I am OK to have debug_plot inside the code at the tricky places but I would avoid this in the signauture

File: src/spikeinterface/core/waveform_extractor.py (hunk 16 above)
Comment (alejoe91) [reply src/spikeinterface/core/waveform_extractor.py:2035]: It is allowed, but only in mem in this case

File: src/spikeinterface/postprocessing/template_metrics.py (hunk 17 above)
Comment (alejoe91) [reply src/spikeinterface/postprocessing/template_metrics.py:224]: Ok, I can remove it and set it as a flag on top of the file

File: src/spikeinterface/core/waveform_extractor.py (hunk 16 above)
Comment (samuelgarcia) [reply src/spikeinterface/core/waveform_extractor.py:2035]: henestly set_params to a read only should be not aalowed.
What is the use here ?

Comment (alejoe91) [reply src/spikeinterface/core/waveform_extractor.py:2035]: It skips saving to disk!

**Review by samuelgarcia — APPROVED** [review]

**Pull request merged**